src/ekka_rlog_agent.erl

```diff
@@ -0,0 +1,160 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc This module implements a gen_statem which pushes rlogs to
+%% a remote node.
+%%
+%% The state machine consists of 3 states:
+%% `catchup', `switchover', `normal', which are explained in detail below:
+%%
+%%
+%% All sends are done as `gen_rpc' calls to the replicant node.
+
+-module(ekka_rlog_agent).
+
+-behaviour(gen_statem).
+
+%% API:
+-export([start_link/3, stop/1]).
+
+%% gen_statem callbacks:
+-export([init/1, terminate/3, code_change/4, callback_mode/0, handle_event/4]).
+
+-include("ekka_rlog.hrl").
+-include_lib("snabbkaffe/include/trace.hrl").
+
+%% Define macros for each state to prevent typos:
+-define(catchup, catchup).
+-define(switchover, switchover).
+-define(normal, normal).
+
+-type state() :: ?catchup | ?switchover | ?normal.
+
+-record(d,
+        { shard                :: ekka_rlog:shard()
+        , subscriber           :: ekka_rlog_lib:subscriber()
+        , seqno          = 0   :: integer()
+        }).
+
+-type data() :: #d{}.
+
+-type fsm_result() :: gen_statem:event_handler_result(state()).
+
+%%--------------------------------------------------------------------
+%% API functions
+%%--------------------------------------------------------------------
+
+start_link(Shard, Subscriber, ReplaySince) ->
+    gen_statem:start_link(?MODULE, {Shard, Subscriber, ReplaySince}, []).
+
+stop(Pid) ->
+    try
+        gen_statem:call(Pid, stop, infinity)
+    catch
+        exit : {noproc, _} ->
+            %% race condition, the process exited
+            %% before or during this call
+            ok
+    end.
+
+%%--------------------------------------------------------------------
+%% gen_statem callbacks
+%%--------------------------------------------------------------------
+
+callback_mode() -> [handle_event_function, state_enter].
+
+-spec init({ekka_rlog:shard(), ekka_rlog_lib:subscriber(), ekka_rlog_lib:txid()}) -> {ok, state(), data()}.
+init({Shard, Subscriber, ReplaySince}) ->
+    logger:update_process_metadata(#{ domain     => [ekka, rlog, agent]
+                                    , shard      => Shard
+                                    , subscriber => Subscriber
+                                    }),
+    D = #d{ shard          = Shard
+          , subscriber     = Subscriber
+          },
+    %% TMP workaround until replaying from the old logs is figured out:
+    subscribe_realtime(D),
+    {ok, ?normal, D}.
+
+-spec handle_event(gen_statem:event_type(), _EventContent, state(), data()) ->
+          gen_statem:event_handler_result(state()).
+%% Events specific to `?normal' state:
+%% TODO: Shouldn't we handle the other mnesia events here (i.e., delete and delete_object)?
```

Comment:
I'm responsible for this comment, methinks, and it is wrong. I didn't understand that the subscription is to the "fake" shard table, and that this will only generate write operations.

Reply:
Actually, it might be a good idea to include this in the technical description. The whole thing with the shard table, dev/null backend. Perhaps it is there already, and I'm not finding it.

Reply:
Good suggestion. No, the null backend is currently undocumented.

Reply:
https://github.com/emqx/ekka/pull/100